### ISSUE TITLE
Word2vec and cython

### ISSUE THREAD

**Issue by djstrong** [piskvorky/gensim#285]
I tried to compile word2vec training code using cython, but after importing it stops execution while training model:
```
...
2015-01-22 23:19:02,192 : INFO : PROGRESS: at sentence #410000, processed 2993042 words and 336718 word types
2015-01-22 23:19:02,240 : INFO : PROGRESS: at sentence #420000, processed 3067635 words and 342330 word types
2015-01-22 23:19:02,261 : INFO : collected 344559 word types from a corpus of 3100055 words and 424434 sentences
2015-01-22 23:19:02,380 : INFO : total 56455 word types after removing those with count<5
2015-01-22 23:19:02,380 : INFO : constructing a huffman tree from 56455 words
2015-01-22 23:19:04,667 : INFO : built huffman tree with maximum node depth 19
2015-01-22 23:19:04,710 : INFO : resetting layer weights
2015-01-22 23:19:05,592 : INFO : training model with 1 workers on 56455 vocabulary and 100 features, using 'skipgram'=1 'hierarchical softmax'=1 'subsample'=0 and 'negative sampling'=0
2015-01-22 23:19:05,616 : DEBUG : putting job #0 in the queue, qsize=0
2015-01-22 23:19:05,616 : DEBUG : putting job #1 in the queue, qsize=1
2015-01-22 23:19:05,617 : DEBUG : putting job #2 in the queue, qsize=2
```
It looks like jobs aren't executed by worker.
python 2.7.6
cython 0.21.3
gensim 0.10.3

**Comment by djstrong** [comment]
The same with cython 0.21.2 and gensim 0.10.2 on other computer with Ubuntu (also).

**Comment by piskvorky** [comment]
What do you mean, tried compiling with Cython?

Word2vec is compiled automatically, from C (not Cython), when you install gensim.

**Comment by djstrong** [comment]
I mean compiling my training code - it is from kaggle tutorial. This code is working in python, but not in cython.

**Comment by djstrong** [comment]
Ok, sorry for bother you. I thought I have to compile my code with cython to run training on all cores, but it works without explicit running cython. So this bug(?) is not necessary to fix.

**Comment by piskvorky** [comment]
I'm still a little confused about what is being compiled, where.

Feel free to close if you've solved the issue, or give more info if not :)

**Comment by djstrong** [comment]
Compiled is my `train.py` to `train.so`.
train.py:
```
import gensim
import os
import  logging
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

class MySentences(object):
     def __init__(self, dirname):
         self.dirname = dirname
 
     def __iter__(self):
         for fname in os.listdir(self.dirname):
             for line in open(os.path.join(self.dirname, fname)):
                 yield line.split()

sentences = MySentences('corpus')
model = gensim.models.Word2Vec(min_count=1, workers=4)
model.build_vocab(sentences)
model.train(sentences) 
```
setup.py:
```
from distutils.core import setup
from Cython.Build import cythonize

setup(
  name = 'Hello world app',
  ext_modules = cythonize("train.py"),
)
```
`python setup.py build_ext --inplace`
```
python
>>> import train
```

**Comment by piskvorky** [comment]
1) You can simply give `sentences` into the constructor of Word2Vec (it will call `build_vocab` and `train` for your automatically, no need for these two extra lines)
2) I don't think Cythonizing this gives you much -- did you profile? I'd expect the bottleneck would be training, which is already done in C, not your code for iterating sentences. What is the difference in speed when you run train.py directly, without Cython?
3) If I understand correctly, the problem is that the word2vec training hangs, correct?

**Comment by djstrong** [comment]
I thought it was necessary to cythonize to run in parallel, but it is not.
Yes, word2vec training hangs - you can close if you don't want fixing that.

**Comment by piskvorky** [comment]
If it hangs I want to fix it -- but I need to know if the problem is inside gensim, or in your `train.py` cython script.

Does it hang when you run `train.py` directly, with cython?

**Comment by djstrong** [comment]
No, `$ python train.py` works normally.

**Comment by piskvorky** [comment]
Ok, so I'm closing this -- probably related to cython somehow, I'm not sure what could be causing this.

**Comment by djstrong** [comment]
Thanks!
